Client/Frontend/Settings/SearchCustomEngineViewController.swift

```diff
@@ -0,0 +1,727 @@
+              let url = URL(string: urlText),
+              url.isWebPage()  else {
+            completion(nil, SearchEngineError.invalidQuery)
+            return
+        }
+    
+        // Check Engine Exists
+        guard profile.searchEngines.orderedEngines.filter({ $0.shortName == name }).isEmpty else {
+            completion(nil, SearchEngineError.duplicate)
+            return
+        }
+
+        var engineImage = #imageLiteral(resourceName: "defaultFavicon")
+
+        guard let hostUrl = host else {
+            let engine = OpenSearchEngine(shortName: name, image: engineImage, searchTemplate: template, isCustomEngine: true)
+
+            completion(engine, nil)
+            return
+        }
+
+        fetcher = FaviconFetcher(siteURL: hostUrl, kind: .favicon)
+
+        fetcher?.load { siteUrl, attributes in
+            if let image = attributes.image {
+                engineImage = image
+            }
+
+            let engine = OpenSearchEngine(shortName: name, image: engineImage, searchTemplate: template, isCustomEngine: true)
+
+            completion(engine, nil)
+        }
+    }
+    
+    private func getSearchTemplate(with query: String) -> String? {
+        let searchTermPlaceholder = "%s"
+        let searchTemplatePlaceholder = "{searchTerms}"
+        
+        if query.contains(searchTermPlaceholder) {
+            return query.replacingOccurrences(of: searchTermPlaceholder, with: searchTemplatePlaceholder)
+        }
+        
+        return nil
+    }
+    
+    private func checkManualAddExists() {
+        guard let url = urlText, let title = titleText  else {
+            return
+        }
+        
+        if !url.isEmpty, !title.isEmpty {
+            changeAddButton(for: .enabled)
+        } else {
+            changeAddButton(for: .disabled)
+        }
+    }
+}
+
+// MARK: - UITextViewDelegate
+
+extension SearchCustomEngineViewController: UITextViewDelegate {
+    
+    func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
+        guard text.rangeOfCharacter(from: .newlines) == nil else {
+            textView.resignFirstResponder()
+            return false
+        }
+
+        return textView.text.count + (text.count - range.length) <= Constants.urlEntryMaxCharacterCount
+    }
+    
+    func textViewDidChange(_ textView: UITextView) {
+        changeAddButton(for: .disabled)
+        
+        urlText = textView.text
+
+        if let text = textView.text?.trimmingCharacters(in: .whitespacesAndNewlines),
+           let encodedText = text.addingPercentEncoding(withAllowedCharacters: .urlFragmentAllowed),
+           let url = URL(string: encodedText),
+           url.host != nil,
+           url.isWebPage() {
+            if let scheme = url.scheme, let host = url.host {
+                self.host = URL(string: "\(scheme)://\(host)")
+            }
+        }
```

Comment:
We should be debouncing this (~0.5s should be fine) to ensure we don't make unnecessary requests for every character typed

Reply:
Fix at https://github.com/brave/brave-ios/pull/3231/commits/0771991d6e130c9beb8e7ddc43114ba55913c249